Add C functions which are declared in headers

bjfish opened the pull request:
I used this tool to find the missing functions:
https://gist.github.com/bjfish/9df9182be1dc670a4be4f5c1949cd418

Then, I made some updates to make things compile correctly.

eregon commented:
Interesting.
I have two concerns:
* This file is already huge, and adding ~1800 lines is not helping, should we split based on the category like alloc.c/string.c/array.c/etc?
* What is the gain to have these functions defined but not implemented? Allow to compile more C-exts which use `try_link`/`have_func`/etc? But then it's quite likely they won't run, isn't it? (e.g. if they check for `rb_profile_frames` most likely the C-ext will use that function).

chrisseaton commented:
> What is the gain to have these functions defined but not implemented?

It means we can forget about Sulong link errors - if someone tries to call a function we haven't implemented we get a nice simple Ruby exception for it.

Also when I need to implement a function I can just go straight to the stub-implementation and work on it - no need to find the declaration and copy it across.

I really like this change.

chrisseaton commented:
What is the plan for when we update Ruby?

eregon commented:
I see, that's indeed great advantages :)
@chrisseaton Any thoughts about splitting the huge ruby.c file or you think it does not matter?
It doesn't need to be part of this PR, but we better do it sooner than later if we decide to do it.

chrisseaton commented:
> Any thoughts about splitting the huge ruby.c file or you think it does not matter?

I don't think it matters but I wouldn't stop anyone else doing it. It's 1800 lines I don't think that's excessive, and the vast majority just call back into Ruby.

eregon commented:
I mean it's 4800 lines long now in total.

eregon commented:
Could you add the tool under `tool` and clean it a bit?
Then we can use it when we update Ruby and in case there is a bug in the generated code more easily re-run it.

bjfish commented:
@eregon @chrisseaton I've cleaned up the script and added it to the `tool` folder.

I think it would nice to organize things in multiple C files someday.

eregon commented on tool/add_missing_c.rb:
```diff
@@ -0,0 +1,96 @@
+          ]
+
+header_files.each do |f|
+    File.open(f).each_line do |line|
```
`File.foreach` is better here, as it doesn't leak a file descriptor.

bjfish replied:
Fixed.

eregon commented on tool/add_missing_c.rb:
```diff
@@ -0,0 +1,96 @@
+  end
+  definition = ruby_c_definitions[m][:full]
+  definition.gsub!(/[\n\t]/,' ')
+  definition = definition.split(' ').join(' ')
```
Minor: `.squeeze(" ")` might be nice here instead of split+join.

bjfish replied:
Fixed.
